sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/UnwrapCastInBinaryComparison.scala

```diff
@@ -155,7 +155,16 @@ object UnwrapCastInBinaryComparison extends Rule[LogicalPlan] {
       list.foreach {
         case lit @ Literal(null, _) => nullList += lit
         case lit @ NonNullLiteral(_, _) =>
-          unwrapCast(EqualTo(in.value, lit)) match {
+          val originalEqualTo = EqualTo(in.value, lit)
+          unwrapCast(originalEqualTo) match {
+            // the function `unwrapCast` may returns original expression when the literal can not
+            // cast to fromType, for instance: (the boundreference is of type DECIMAL(5,2))
+            //     CAST(boundreference() AS DECIMAL(10,4)) = 123456.1234BD
+            // Due to `cast(lit, fromExp.dataType) == null` we simply return
+            // `falseIfNotNull(fromExp)`.
+            case equalTo: EqualTo if equalTo == originalEqualTo &&
+              Cast(lit, fromExp.dataType).eval() == null =>
```

Comment:
will there be any cases where `Cast(lit, fromExp.dataType).eval()` is not equal to `null`?

Reply:
if `Cast(lit, fromExp.dataType).eval()` is not equal to `null`, then the return expression of `unwrapCast` should be either `EqualTo(fromExp, lit)` or `falseIfNotNull(fromExp)` which was handled by L168-L169.

https://github.com/apache/spark/blob/5a71a7f7b5c1762677ddbfe39a7c35d645c25e94/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/UnwrapCastInBinaryComparison.scala#L305-L346

> we may have a return value from L322, L331 or L340.